File: src/vct/col/ast/Lemma.java
package vct.col.ast;

/**
 * This class represents magic wand proofs, a.k.a. create blocks.
 * 
 * @author Stefan Blom
 *
 */
public class Lemma extends ASTNode {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

	public final BlockStatement block;
	
	public Lemma(BlockStatement block) {
		this.block=block;
	}

	  
	  @Override
	  public <T> void accept_simple(ASTVisitor<T> visitor){
	    try {
	      visitor.visit(this);
	    } catch (Throwable t){
	      if (thrown.get()!=t){
	        System.err.printf("Triggered by %s:%n",getOrigin());
	        thrown.set(t);
      }
	      throw t;
	    }
	  }
	  
	  @Override
	  public <T> T accept_simple(ASTMapping<T> map){
	    try {
	      return map.map(this);
	    } catch (Throwable t){
	      if (thrown.get()!=t){
	        System.err.printf("Triggered by %s:%n",getOrigin());
	        thrown.set(t);
	      }
	      throw t;
	    }
	  }
	 

}

File: vercors/src/main/java/vct/col/ast/Lemma.scala
package vct.col.ast

import vct.col.util.VisitorHelper

/**
 * This class represents magic wand proofs, a.k.a. create blocks.
 * 
 * @author Stefan Blom, Wytse Oortwijn
 * @param block The block representing the lemma (for the magic wand proof).
 */
class Lemma(val block:BlockStatement) extends ASTNode with VisitorHelper {
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
}